server/mmauth/mmauth.go

```diff
@@ -0,0 +1,249 @@
+
+	redirectCookie := http.Cookie{Name: "oauthredirect", Path: "/", Value: redirectUrl, Expires: expiration, HttpOnly: true}
+	http.SetCookie(w, &redirectCookie)
+
+	oauthState := generateSecurityToken()
+	stateCookie := http.Cookie{Name: "oauthstate", Path: "/", Value: oauthState, Expires: expiration, HttpOnly: true}
+	http.SetCookie(w, &stateCookie)
+
+	authUrl := a.config.AuthCodeURL(oauthState)
+	http.Redirect(w, r, authUrl, http.StatusTemporaryRedirect)
+}
+
+func generateSecurityToken() string {
+	b := make([]byte, 16)
+	rand.Read(b)
+	securityToken := base64.URLEncoding.EncodeToString(b)
+	return securityToken
+}
+
+func deleteCookie(name string, w http.ResponseWriter) {
+	redirectCookie := http.Cookie{Name: name, Path: "/", Value: "", MaxAge: -1, Expires: time.Now().Add(-100 * time.Hour), HttpOnly: true}
+	http.SetCookie(w, &redirectCookie)
+}
+
+func (a *MMAuth) handleOAuthCallback(w http.ResponseWriter, r *http.Request) {
+	// Read oauthState from Cookie
+	stateCookie, err := r.Cookie("oauthstate")
+	if err != nil {
+		a.handleError("oauthstate cookie not found", err, w, r)
+		return
+	}
+
+	redirectCookie, err := r.Cookie("oauthredirect")
+	if err != nil {
+		a.handleError("oauthredirect cookie not found", err, w, r)
+		return
+	}
+
+	// Delete cookies
+	deleteCookie("oauthredirect", w)
+	deleteCookie("oauthstate", w)
+
+	if r.FormValue("state") != stateCookie.Value {
+		log.Println("invalid oauth state")
+		http.Redirect(w, r, "/", http.StatusTemporaryRedirect)
+		return
+	}
+
+	code := r.FormValue("code")
+	token, err := a.config.Exchange(context.Background(), code)
+	if err != nil {
+		a.handleError("code exchange error", err, w, r)
+		return
+	}
+
+	// Get user info from Mattermost
+	client := mattermost.NewAPIv4Client(a.params.MattermostURL)
+	client.SetOAuthToken(token.AccessToken)
+
+	etag := ""
+	user, resp := client.GetMe(etag)
+	if resp.Error != nil {
+		a.handleError("failed getting Mattermost user", resp.Error, w, r)
+		return
+	}
+
+	log.Printf("UserInfo: %v\n", user)
+
+	teams, resp := client.GetAllTeams(etag, 0, 10)
+	if resp.Error != nil {
+		a.handleError("failed getting Mattermost teams", err, w, r)
+		return
+	}
+
+	if len(teams) < 1 {
+		a.handleError("User has no teams", err, w, r)
+		return
+	}
+
+	log.Printf("%d team(s).", len(teams))
+	teamId := teams[0].Id
+
+	channels, resp := client.GetChannelsForTeamForUser(teamId, user.Id, false, etag)
+	if resp.Error != nil {
+		a.handleError("failed getting Mattermost channels for user", err, w, r)
+		return
+	}
+
+	if len(channels) < 1 {
+		a.handleError("User has no channels", err, w, r)
+		return
+	}
+
+	log.Printf("%d channel(s).", len(channels))
+	channelId := channels[0].Id
+
+	log.Printf("userId: %s, teamId: %s, channelId: %s", user.Id, teamId, channelId)
+
+	// Create native user
+	// TODO: Update user info with any changes?
+	nativeUser, _ := a.store.GetUserById(user.Id)
+	if nativeUser == nil {
+		err = a.store.CreateUser(&model.User{
+			ID:          user.Id,
```

Comment:
Directly mapping the focalboard user ID to the MM user ID is an interesting method. In the MM server we use the AuthData field for this and the user would get a unique focalboard ID too. This will work but it will tightly couple the user to the oauth service provider (in this case MM). If you ever wanted to change the auth service from MM to Google for example then this would become problematic since the focalboard user ID would need to change. Same problem if you wanted to support multiple login methods for one focalboard user.

Reply:
Chatted with Chen in a DM about it. Leaving for now, may revisit in the future.